File: packages/julia/src/Seguranca/Ataques/AtaqueChavesFracas.jl
"""
Busca dirigida de chaves fracas: chaves degeneradas/estruturadas não devem
produzir keystream anômalo (repetição, período curto, viés de bits ou
distribuição de bytes distorcida). Complementa o ataque de chaves degeneradas,
que só checa a saída do encrypt.
"""
struct AtaqueChavesFracas <: Ataque
    tamanho::Int
    tolerancia_bits::Float64
end

AtaqueChavesFracas(; tamanho::Integer = 2048, tolerancia_bits::Real = 0.05) =
    AtaqueChavesFracas(Int(tamanho), Float64(tolerancia_bits))

nome(::AtaqueChavesFracas) = "Chaves fracas (busca dirigida)"

function executar(a::AtaqueChavesFracas, alvo::AlvoCriptografico)
    len = length(codeunits(chave_de_teste(alvo)))

    casos = Pair{String,Vector{UInt8}}[
        "zeros" => zeros(UInt8, len),
        "0xFF" => fill(0xff, len),
        "alternada AA/55" => UInt8[i % 2 == 0 ? 0xaa : 0x55 for i in 0:(len-1)],
        "incremental" => UInt8[i & 0xff for i in 0:(len-1)],
        "um bit" => UInt8[i == 0 ? 1 : 0 for i in 0:(len-1)],
        "byte repetido 0x01" => fill(0x01, len),
        "padrão AB" => UInt8[i % 2 == 0 ? 0x41 : 0x42 for i in 0:(len-1)],
    ]

    iv_fixo = zeros(UInt8, tamanho_iv(alvo))
    anomalias = String[]

    for (nome_caso, chave) in casos
        ks = gerar_keystream_bruto(alvo, chave, iv_fixo, "enc", a.tamanho)

        blocos = Set{String}()
        repetidos = 0
        for i in 0:32:(length(ks)-32)
            hex = bytes2hex(ks[i+1:i+32])
            if hex in blocos
                repetidos += 1
            else
                push!(blocos, hex)
            end
        end

        fracao_uns = contar_bits_buffer(ks) / (length(ks) * 8)
        desvio_bits = abs(fracao_uns - 0.5)

        if repetidos > 0
            push!(anomalias, "$nome_caso: $repetidos bloco(s) repetido(s)")
        end
        if desvio_bits > a.tolerancia_bits
            push!(anomalias, "$nome_caso: viés de bits $(fmt(fracao_uns * 100, 1))%")
        end
    end

    vulneravel = !isempty(anomalias)

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "alta" : "info",
        vulneravel ?
        "Anomalias: $(join(anomalias[1:min(5, length(anomalias))], "; "))" :
        "Nenhuma das $(length(casos)) chaves fracas produziu keystream anômalo ($(a.tamanho) bytes cada)",
        Dict{String,Any}("anomalias" => anomalias),
    )
end

File: packages/python/src/Seguranca/Ataques/AtaqueChavesFracas.py
"""Busca dirigida de chaves fracas: chaves degeneradas/estruturadas não devem
produzir keystream anômalo (repetição, período curto, viés de bits ou
distribuição de bytes distorcida). Complementa o ataque de chaves
degeneradas, que só checa a saída do encrypt.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import contar_bits_buffer


class AtaqueChavesFracas:
    def __init__(self, tamanho: int = 2048, tolerancia_bits: float = 0.05):
        self.tamanho = tamanho
        self.tolerancia_bits = tolerancia_bits

    def nome(self) -> str:
        return "Chaves fracas (busca dirigida)"

    def executar(self, alvo) -> ResultadoAtaque:
        tamanho = len(alvo.chave_de_teste().encode())
        casos = [
            ("zeros", bytes(tamanho)),
            ("0xFF", bytes([0xFF] * tamanho)),
            ("alternada AA/55", bytes(0xAA if i % 2 == 0 else 0x55 for i in range(tamanho))),
            ("incremental", bytes(i & 0xFF for i in range(tamanho))),
            ("um bit", bytes(1 if i == 0 else 0 for i in range(tamanho))),
            ("byte repetido 0x01", bytes([0x01] * tamanho)),
            ("padrão AB", bytes(0x41 if i % 2 == 0 else 0x42 for i in range(tamanho))),
        ]

        iv_fixo = bytes(alvo.tamanho_iv())
        anomalias = []

        for nome_caso, chave in casos:
            ks = alvo.gerar_keystream_bruto(chave, iv_fixo, "enc", self.tamanho)

            blocos = set()
            repetidos = 0
            for i in range(0, len(ks) - 32 + 1, 32):
                hex_ = ks[i : i + 32].hex()
                if hex_ in blocos:
                    repetidos += 1
                else:
                    blocos.add(hex_)

            fracao_uns = contar_bits_buffer(ks) / (len(ks) * 8)
            desvio_bits = abs(fracao_uns - 0.5)

            if repetidos > 0:
                anomalias.append(f"{nome_caso}: {repetidos} bloco(s) repetido(s)")
            if desvio_bits > self.tolerancia_bits:
                anomalias.append(f"{nome_caso}: viés de bits {fracao_uns * 100:.1f}%")

        vulneravel = len(anomalias) > 0

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "alta" if vulneravel else "info",
            (
                f"Anomalias: {'; '.join(anomalias[:5])}"
                if vulneravel
                else f"Nenhuma das {len(casos)} chaves fracas produziu keystream anômalo ({self.tamanho} bytes cada)"
            ),
            {"anomalias": anomalias},
        )
